### PR TITLE
Add support for RP-Initiated logout

### LINKED ISSUE #24: Single sign-off: RP-Initiated logout
In #22 are described the requirements of proper session management when dealing with single sign-on providers, and single sign-off is among the others a sensible one as it involves different scenarios; for example:

1. the currently authenticated user in the app clicks on the logout button, which sessions should be invalidated?
2. the user invalidates the single sign-on session from elsewhere (outside Vaadin, or from another Vaadin app), how the current app would know about this?

The first scenario can be managed within the app, by invalidating both Vaadin's and the servlet's sessions and signaling the provider redirecting the user to the provider's own logout URL.

The OpenID Connect standard specify a standardized way to approach the second scenario, which is called [Back-Channel Logout](https://openid.net/specs/openid-connect-backchannel-1_0.html). The spec requires a in-app stateless URL to send logout requests from the provider, with the Session-ID (SID) of the user sessions which should be terminated (being that currently active or not).

Currently Spring does not offer that support, but there is an open request to have this part of the Spring's OIDC implementation: https://github.com/spring-projects/spring-security/issues/7845

We can start implementing support for the first scenario and in the meanwhile investigate if Spring is going to add support for back-channel logout in the short-term.

### PR DESCRIPTION
**Pull request by tamasmak**
## Description

- Add logout method and its implementaion to the `VaadinAuthContext` to handle the single sign-off
- Add logout redirect url to the properties
- Modify the logout button in the Profile view

Fixes #24

Missing:
- [ ] Logout redirect url handling
- [ ] Finalize documentation
- [ ] Tests

Type of change
- [ ] Bugfix
- [x] Feature

### CODE REVIEW

File: auth-starter/src/main/java/com/vaadin/auth/starter/VaadinAuthContextImpl.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    @Override
+    public Optional<OidcUser> getAuthenticatedUser() {
+        return Optional.of(SecurityContextHolder.getContext())
+                .map(SecurityContext::getAuthentication)
+                .map(Authentication::getPrincipal)
+                .filter(OidcUser.class::isInstance).map(OidcUser.class::cast);
+    }
+
+    @Override
+    public void logout() {
```
Comment (knoobie) [inline auth-starter/src/main/java/com/vaadin/auth/starter/VaadinAuthContextImpl.java:48]: Does this solution tackle the problem described by @heruan `But this does not take into account any custom logout handler that could be added to the security configuration, so it's not optimal.`? If not, it probably has to be documented and if feasible Vaadin should provide it's own hooking mechanism to allow people to hook into the logout

Comment (heruan) [reply auth-starter/src/main/java/com/vaadin/auth/starter/VaadinAuthContextImpl.java:48]: At this point it does not, but we already have prototyped a proper alternative that will be replaced before this gets merged: all the handlers (Spring and any custom one) will be collected right before the security chain is built, so they'll be used in this method.

**Comment by heruan** [comment]
There was some code shared with another branch which led to conflicts when merging with main, I'm converting this to draft until they're fixed.

**Comment by heruan** [comment]
The PR is now ready for review. On top of the previous commits, it now has:
- [x] A default implementation of `VaadinAuthContext::logout` which takes care of logout handlers
- [x] Auto-configuration for logout and exception handling
- [x] Inline comments to explain each step of auto-configuration
- [x] JavaDoc on new methods and classes
- [x] Tests to ensure logout handlers are set and executed

**Comment by knoobie** [comment]
> Inline comments to explain each step of auto-configuration

Is it planned to let this module stay open source? Or is it planned to change it later to closed source like done with the addition of the legacy components? I can totally understand the reasoning behind the license / payed model. But I personally think that security related code should stay open source to find and report security problems instead of forcing people to decompile it to find bugs or security holes. This comment comes from the fact that inline comments are used, which in turn are missing if close source is used and the source files aren't distributed.

**Comment by heruan** [comment]
> Is it planned to let this module stay open source? Or is it planned to change it later to closed source like done with the addition of the legacy components?

Hey @knoobie, thanks for the question! Current plan is to keep the sources open, although using the add-on or parts of the code will require a commercial license.

**Comment by knoobie** [comment]
> the add-on or parts of the code will require a commercial license

That is totally valid! Thanks for the feedback, really appreciated!

File: auth-demo/src/main/resources/application.yml [hunk 2]
```diff
@@ -3,3 +3,4 @@ server:
 vaadin:
   auth:
     auto-configure: true
+    logout-redirect-url: /
```
Comment (DiegoCardoso) [inline auth-demo/src/main/resources/application.yml:6]: It should be `logout-redirect-route`, according to the getters/setters at `VaadinAuthProperties` class.
```suggestion
    logout-redirect-route: /
```

**Review by DiegoCardoso — CHANGES_REQUESTED** [review]
Initial comments.

**Review by DiegoCardoso — APPROVED** [review]
The changes look good to me. 

I was able to configure and see that the user's session is correctly removed when I was using it with Keycloak.
With Okta, I wasn't able to make it work, but because it fails with a CORS error when trying to reach the provider's API. I tried to configure it there, but not successfully. Judging by the format of the request, I believe it should work when properly configured.

**Pull request merged**